File: Assets/Scripts/Networking/Message/CloseConnectionMessage.cs
﻿using System;
using Networking.Message.Utils;
using Utils.Extensions;

namespace Networking.Message
{
    /// <summary>
    /// Сообщение о закрытии клиента
    /// </summary>
    public class CloseConnectionMessage: IMessage
    {
        /// <summary>
        /// Тип сообщения
        /// </summary>
        public MessageType MessageType => MessageType.CloseConnection;

        /// <summary>
        /// Расположен ли сокет на стороне клиента
        /// </summary>
        public readonly bool SendGoodbyeMessage;

        public CloseConnectionMessage(bool sendGoodbyeMessage)
        {
            SendGoodbyeMessage = sendGoodbyeMessage;
        }
        
        /// <summary>
        /// Сериализует сообщение в массив байтов 
        /// </summary>
        public byte[] Serialize()
        {
            //MessageType + SendGoodbyeMessage
            const ushort length = (ushort) (1 + sizeof(bool));
            this.CreateMessage(length, out var data);
            
            var offset = MessageExtensions.HEADER_LENGTH;
            offset = MessageExtensions.SetByte(in data, (byte) MessageType, offset);
            MessageExtensions.SetBytes(in data, SendGoodbyeMessage, offset);
            return data;
        }

        /// <summary>
        /// Обрабатывает вызов десериализации сообщения 
        /// </summary>
        public static IMessage Deserialize(in byte[] data)
        {
            var offset = MessageExtensions.HEADER_LENGTH + 1; 
            return new CloseConnectionMessage(MessageExtensions.GetBoolean(data, ref offset));
        }
    }
}

File: Python/Socket/Messages/CloseConnectionMessage.py
import Socket.Messages.Utils.Params as Params
import Socket.Utils.ByteConverter as ByteConverter

from Socket.Messages.Message import Message
from Socket.Messages.Utils.MessageTypes import MessageTypes


# Класс сообщения о закрытии сокета
class CloseConnectionMessage(Message):
    SendGoodbyeMessage: bool

    def __init__(self, sendGoodbyeMessage):
        length = Params.MESSAGE_HEADER_LENGTH + 1 + 1
        self.MessageLength = length
        self.MessageType = MessageTypes.CloseConnection
        self.SendGoodbyeMessage = sendGoodbyeMessage

    def Serialize(self):
        returnArray = bytearray()
        returnArray.extend(ByteConverter.GetBytesFromInteger(self.MessageLength))
        returnArray.extend(ByteConverter.GetBytesFromByteInteger(self.MessageType))
        returnArray.extend(ByteConverter.GetBytesFromBool(self.SendGoodbyeMessage))
        return returnArray

    @staticmethod
    def Deserialize(data: bytearray):
        offset = Params.MESSAGE_HEADER_LENGTH + 1
        return CloseConnectionMessage(ByteConverter.GetBool(data, offset=offset))
